### DISCUSSION TITLE
Supporting default translation values

### DISCUSSION THREAD

**Discussion by halilb** [ivanhofer/typesafe-i18n#453] [Ideas]
Thanks for the great library @ivanhofer!

My colleague told me about Elastic's [kbn-i18n](https://github.com/elastic/kibana/tree/main/packages/kbn-i18n) library while I was integrating this one into our project.

It has one requirement to keep the default English version right next to the template where the translation is used. This is because they want to fuzzy-find English texts in the project and jump right at where it's used. It also adds more context to template files so devs would know better what they're looking at.

> One of our technical requirements is to have default messages in the templates themselves, and those messages will always be in English, so we don't have to keep en.json file in repository. We can generate that file from defaultMessages defined inline.

It comes with a few trade-offs so we're not sure if using default values is worth the effort. So I wanted to ask you before going forward with this decision: can you think of a way to build it as a wrapper of `typesafe-i18n`?

**Comment by ivanhofer** [comment]
Hi @halilb,

I'm not sure if I understood everything correctly. I try to summarize it:

1. keep translations near the component that uses the translation
   `typesafe-i18n` uses the opposite approach and stores translations in a central place (`./src/i18n/[locale]/...`). You could still go with the other approach but it is a bit of a hack. Since `typesafe-i18n` builds on top of usual `TypeScript` files, you can do the following:
   
   store translations near your components
   _/src/components/login/translations.ts_
   ```ts
   export const en = {
      usernameLabel: 'username:',
      passwordLabel: 'password:',
      cta: 'login'
   }

   export const de= {
      usernameLabel: 'Benutzername:',
      passwordLabel: 'Passwort:',
      cta: 'Einloggen'
   }
   ```
   and import them in your dictionary
   _/src/i18n/en/index.ts_
   ```ts
   import type { BaseTranslation } from '../i18n-types'
   import { en as login } from '../../components/login/translations'
   
   const en: BaseTranslation = {
      // some base translations
      continue: 'Continue',
      // component-specific translations
      components: {
         login,
         // ... other components
      }
   }
   export default en
   ```
2. use the base translation as a key
   Most libraries suggest to organize your translations via keys
   ```tsx
   {
      'hello': 'Hello {name}!'
   }

   <h1>{LL.hello({ name: 'John' })}</h1>
   ```
   instead if inlineing them
   ```tsx
   const name = 'John'

   <h1>{LL`Hello ${name}!`}</h1>
   ```
   There is a middle-way you could use with `typesafe-i18n` (and most other libraries):
   Use the message itself as a key.
   ```html
   <h1>{LL['Hello {name}!']({ name: 'John' })}</h1>
   ```
   But you still would need to add it to the base translations
   ```ts
   {
      'Hello {name}!': 'Hello {name}!'
   }
   ```

I'm not sure if this answers your questions 😅
Let me know if something is still unclear.

**Reply by halilb** [reply]
Thanks for the quick reply @ivanhofer! I forgot to add an example that could have made it clearer.

```ts
// src/i18n/en/index.ts
export const en = {
   // Not repeated here as it's defined with the `defaultText` value
   hello: '{defaultText}',
}

// src/i18n/de/index.ts
export const de= {
  helllo: 'Hallo {name}. Wie gehts?',
}

// src/some_component.ts
const { LL } = useI18nContext();
const helloText = LL.hello({
  defaultText: 'Hello {name}. How are you?',
  name: 'John',
});

// Should print 'Hello John. How are you?' for English
console.log(helloText);
```

The benefit here is I'll jump right at `some_component` when I fuzzy-search `How are you?` in my codebase.

Trade-offs are:

* We'll always load English translations into the client so it might increase the network traffic for non-English users.
* It gets harder to provide a single translation file in English to translators.

We're also trying to understand if we're overengineering the problem here so I'd love to hear your opinions.

**Reply by ivanhofer** [reply]
Ah, now I got it. Thanks for the example.

I can understand that jumping to the actual text directly inside the component would increase DX a lot when searching for text snippets.

The example above is not possible with `typesafe-i18n` because the `generator` needs to parse the base translation file (en) in order to create the `i18n-types.ts` file. 

Loading always the english translation on top of the actual locale is far from ideal, but would probably not really affect the performance of your application.

In theory you could write a `vite` (or other built-tool) plugin that extracts all `defaultText`s and writes them to the base translation file. But I'm not sure if it is worth the effort.

**Comment by halilb** [comment]
I just realized there is a VSCode plugin that solves some of the pain points I described: https://github.com/lokalise/i18n-ally We are using `neovim` and there is no such library for that, unfortunately.

One thing that'd help us could be removing the translation from `RooTranslation` types so we can jump into `TranslationFunctions` when we fuzzy-find a text.

Example output:

```ts
type RootTranslation = {
  server: {
    /**
     * You need to post a comment first to be able to complete this lesson.
     */
    needCommentFirst: string;
  };
};

export type TranslationFunctions = {
  server: {
    /**
     * You need to post a comment first to be able to complete this lesson.
     */
    needCommentFirst: () => LocalizedString;
  };
};
```

Is there a way to modify auto-generated `RootTranslation` types to remove "You need to post..." text there to prevent duplication?

**Reply by ivanhofer** [reply]
> I just realized there is a VSCode plugin that solves some of the pain points I described

`i18n-ally` will just display the translation when you have openend the file in your editor. It will not help you when you are searching for the string.

> Is there a way to modify auto-generated RootTranslation types to remove "You need to post..." text there to prevent duplication?

Currently not. The reason why `RootTranslation` contains the JSDoc coments is: when you are translating an entry in e.g. `i18n/de/index.ts`, hovering over the entry will show you the english translation and you don't need to jump to `i18n/en/index.ts` in order to check it.

But this could be an optional feature that could be disabled by setting an option in the `.typesafe-i18n.json` file.
Are you interested in contributing a PR for that feature?

**Reply by halilb** [reply]
Yes! I'd love to contribute if you could describe to me where the `RootTranslation` is generated. I went over the source code but couldn't exactly identify the location for that.

**Reply by ivanhofer** [reply]
Yeah, the generator is a bit hard to understand at the fisrt glance 😅

1. Here is the part where the JSDoc comment gets added: https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/generator/src/files/generate-types/translations.mts#L61
2. A new option needs to be added to the `output-handler` so the JSDoc comment can be disabled/enabled: https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/generator/src/output-handler.mts
3. This option can be set by passing it to the `GeneratorConfig` https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/config/src/types.mts#L24
4. The readme has to be modified as well to document the new option: https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/generator/README.md#options
5. And a new test should be added: https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/generator/test/generator.test.ts#L276

**Reply by halilb** [reply]
Thanks for the guidance @ivanhofer!

**Reply by ivanhofer** [reply]
I have another Idea how to tackle this:

There exists an invisible character character called "ZeroWidthSpace" (`​`). We could use that character and inject it between each character on the `RootTranslation` type.

I have configured VS Code to show such special characters in red:

![image](https://user-images.githubusercontent.com/21335119/192691297-65f7b7eb-cd8f-4ae4-a495-55d5aa73100a.png)

If you search for "Hello" VS Code will only find it once. But you still would get the same hover-experience when moving the mouse over the key:

![image](https://user-images.githubusercontent.com/21335119/192691775-beaeeb1e-81e5-45c5-973c-af6888a08636.png)

What do you think?

**Reply by halilb** [reply]
Nice! This would do the job for our purposes. Do you consider making it default or putting it behind a configuration option?

**Reply by ivanhofer** [reply]
I think we can make this the default behaviour. I can implement this in the coming days

**Reply by halilb** [reply]
Sounds great! Thank you @ivanhofer!

**Reply by ivanhofer** [reply]
Implemented and released in version `5.14.0` 🎉

**Reply by halilb** [reply]
That's really fast! Thanks a lot, @ivanhofer. I'll try it out.
